server/src/main/java/org/opensearch/gateway/ShardsBatchGatewayAllocator.java

```diff
@@ -0,0 +1,734 @@
+    );
+
+    private final RerouteService rerouteService;
+    private final PrimaryShardBatchAllocator primaryShardBatchAllocator;
+    private final ReplicaShardBatchAllocator replicaShardBatchAllocator;
+    private Set<String> lastSeenEphemeralIds = Collections.emptySet();
+
+    // visible for testing
+    protected final ConcurrentMap<String, ShardsBatch> batchIdToStartedShardBatch = ConcurrentCollections.newConcurrentMap();
+
+    // visible for testing
+    protected final ConcurrentMap<String, ShardsBatch> batchIdToStoreShardBatch = ConcurrentCollections.newConcurrentMap();
+    private final TransportNodesListGatewayStartedShardsBatch batchStartedAction;
+    private final TransportNodesListShardStoreMetadataBatch batchStoreAction;
+
+    @Inject
+    public ShardsBatchGatewayAllocator(
+        RerouteService rerouteService,
+        TransportNodesListGatewayStartedShardsBatch batchStartedAction,
+        TransportNodesListShardStoreMetadataBatch batchStoreAction,
+        Settings settings
+    ) {
+        this.rerouteService = rerouteService;
+        this.primaryShardBatchAllocator = new InternalPrimaryBatchShardAllocator();
+        this.replicaShardBatchAllocator = new InternalReplicaBatchShardAllocator();
+        this.batchStartedAction = batchStartedAction;
+        this.batchStoreAction = batchStoreAction;
+        this.maxBatchSize = GATEWAY_ALLOCATOR_BATCH_SIZE.get(settings);
+    }
+
+    @Override
+    public void cleanCaches() {
+        Releasables.close(
+            batchIdToStartedShardBatch.values().stream().map(shardsBatch -> shardsBatch.asyncBatch).collect(Collectors.toList())
+        );
+        batchIdToStartedShardBatch.clear();
+        Releasables.close(
+            batchIdToStoreShardBatch.values().stream().map(shardsBatch -> shardsBatch.asyncBatch).collect(Collectors.toList())
+        );
+        batchIdToStoreShardBatch.clear();
+    }
+
+    // for tests
+    protected ShardsBatchGatewayAllocator() {
+        this.rerouteService = null;
+        this.batchStartedAction = null;
+        this.primaryShardBatchAllocator = null;
+        this.batchStoreAction = null;
+        this.replicaShardBatchAllocator = null;
+        this.maxBatchSize = DEFAULT_SHARD_BATCH_SIZE;
+    }
+
+    // for tests
+
+    @Override
+    public int getNumberOfInFlightFetches() {
+        int count = 0;
+        for (ShardsBatch batch : batchIdToStartedShardBatch.values()) {
+            count += (batch.getNumberOfInFlightFetches() * batch.getBatchedShards().size());
+        }
+        for (ShardsBatch batch : batchIdToStoreShardBatch.values()) {
+            count += (batch.getNumberOfInFlightFetches() * batch.getBatchedShards().size());
+        }
+
+        return count;
+    }
+
+    @Override
+    public void applyStartedShards(final List<ShardRouting> startedShards, final RoutingAllocation allocation) {
+        for (ShardRouting startedShard : startedShards) {
+            safelyRemoveShardFromBothBatch(startedShard);
```

Comment:
Trying to understand this logic of removing shard from the batch. Ideally the batch should be immutable right as in you had a batch you sent request to all nodes, you got response or any node failed. Then appropriately we handle response. Now the lifetime of batch is over. why do we need to remove entries from batch? will this not complicate the flow?

Reply:
**No new shards can be added in a batch, only shard removal is possible based on STARTED or failures (for retries).**

Clearing the shard from batch help removing the actual shardMetadata content stored in the cache (AsyncShardFetch cache object) before all the shards of the batch are assigned. Right now, I've written the flow similar to older one for every shard. We only used to remove the shard from cache when we get shard-started or failed event. 

We remove shard from batch 2 times
1. Based on shard-started / failed pending task execution.
2. If we get a retry-able exception for a shard during fetch.

Removing failed shards from batch also help, as that clearly gives us a signal that there are no shards remaining in the batch.